src/chrome/utils/localizedError.ts

```diff
@@ -0,0 +1,119 @@
+                if (textKey !== undefined) {
+                    return textKey;
+                }
+            }
+        }
+
+        return `${file}:${key}`;
+    } else if (typeof key === 'string') {
+        return key;
+    } else {
+        return 'getNextLastKey parameter was unexpected';
+    }
+}
+
+/** Normalize the file path to the format used in the localization bundles (Relative path, no extension, and forward slashes) */
+function normalizeFilePath(file: string): string {
+    const fileRelativePath = path.relative(bundleFolderPath, file);
+    const normalizedPath = fileRelativePath.replace(/\\/g, '/').replace(/\.js$/g, '');
+    return normalizedPath;
+}
+
+const originalConfig = nls.config;
+let currentLocale: string | undefined = undefined;
+
+/** We replace nls.config with our own version so we can know which locale is set, and force all the localize functions to be re-created when needed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).config = (opts?: nls.Options) => {
+    currentLocale = _.defaultTo(opts, { locale: undefined }).locale;
+    return originalConfig(opts);
+};
+
+const originalLoadMessageBundle = nls.loadMessageBundle;
+/** We replace nls.loadMessageBundle with our own custom version, which will re-created the localize function after the locale has been changed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).loadMessageBundle = (file?: string) => {
+    let localizeForFile = originalLoadMessageBundle(file);
+    let localizeForFileLocale = currentLocale;
+    return (key: string | LocalizeInfo, message: string, ...args: (string | number | boolean | undefined | null)[]) =>
+        {
+            // Has the locale changed since we created the localize function?
+            if (localizeForFileLocale !== currentLocale) {
+                localizeForFileLocale = currentLocale;
+                // If so, re-create it so it'll use the new local (the localize caches the local, so we do need to re-create it after changing the locale)
+                localizeForFile = originalLoadMessageBundle(file);
+            }
+
+            // Call our custom localize which will store the localization key for the LocaliedErrors
+            return customLocalize(file, localizeForFile, <string>key, message, ...args);
+        };
+};
+
+// Consume the key of the latest call to the localize function
+function consumeLastLocalizationKey(): string {
+    const lastKeyToReturn = lastKeyOfLocalize;
+    lastKeyOfLocalize = defaultLastKeyOfLocalize;
+    return lastKeyToReturn;
+}
+
+/** Error class which automatically extracts the localization key of the latest call to localize (So we can easily deduplicate errors on telemetry) */
+export class LocalizedError extends Error {
+    public readonly errorCode = consumeLastLocalizationKey();
```

Comment:
This is not a safe assumption, how can you know that localize will always be called immediately before creating a LocalizedError?

Reply:
There is no warranty. At the moment all the calls of new LocalizedError( have localize as it's first argument, so it should be true.
If this where ever to be false we would return "last localization key couldn't be determined" and see that in the telemetry, and we could investigate more to figure out why that happened.